### ISSUE TITLE
pnpm up --save-exact doesn't work

### ISSUE THREAD

**Issue by unional** [pnpm/pnpm#6398]
### pnpm version:

8.2.0

### Code to reproduce the issue:

`pnpm up -r --save-exact pkg@1.0.1

### Expected behavior:

it should pin the version to 1.0.1

### Actual behavior:

package.json contains `pkg: '^1.0.1'`

### Additional information:

 - `node -v` prints: 18.15.0
 - Windows, macOS, or Linux?: Windows.

It seems like adding `pkg@^?` has no difference. It always keep whatever was there.
i.e.: 
```
// before
pkg: "^1.0.0"`

// after (even when doing pkg@1.0.1)
pkg: "^1.0.0"

// before
pkg: "1.0.0"

// after (even when doing pkg@*, pkg@^*, pkg@^1.0.1)
pkg: "1.0.1
```

**Comment by 0x80** [comment]
I just ran into this as well. It would be really helpful to get this to work in order to pin exact versions across a monorepo.

I just spent hours trying to work out why my system broke, only to find out that it was firebase-admin 10.10 vs 10.9, with nothing remotely related mentioned in the release notes. Bummer 😓

**Comment by 0x80** [comment]
Maybe related, if you try to install a package with -E that already exists in your manifest, it also doesn't have any affect on the "^" that is already there.

For example:

`pnpm add mobx-react-lite@3.4.3 -E`

I would expect this to store the dependency with exact versioning no matter if it was already installed or not.

**Comment by eckdanny** [comment]
`up [-E|--save-exact] [<pkg>...]` _would_ be helpful. Just had similar experience rev'ing all workspaces to same version of <samp>prettier</samp>. Should be inclusive of `[-L|--latest]`. 

reported against `pnpm@8.15.1`

**Comment by Armadillidiid** [comment]
Just encountered this issue in a monorepo. What I found to work is to uninstall the package and reinstall it with `--save-exact` flag. If the package is already present in your `pnpm-lock.yaml`and you try to reinstall using `--save-exact`, it won't take effect.

**Comment by fabiobaser** [comment]
> Just encountered this issue in a monorepo. What I found to work is to uninstall the package and reinstall it with `--save-exact` flag. If the package is already present in your `pnpm-lock.yaml`and you try to reinstall using `--save-exact`, it won't take effect.

This was my problem as well. Deleting the `pnpm-lock.yaml` in the monorepo root and then installing in my package with `pnpm add -E <pkg>@<version>` it worked as intended. But I don't feel so good trashing my complete `pnpm-lock.yaml` only for this.

**Comment by karlhorky** [comment]
@zkochan @btea I can confirm this issue - I ran into this today with [`pnpm@11.22.0`](https://github.com/pnpm/pnpm/releases/tag/v11.22.0):

```bash
$ mkdir c && cd c

$ pnpm add typescript@6
...
dependencies:
+ typescript 6.0.3 (7.0.2 is available)

Done in 642ms using pnpm v11.22.0
$ cat package.json
{
        "dependencies": {
                "typescript": "^6.0.3"
        }
}

$ pnpm add --save-exact typescript@6.0.3
...
Progress: resolved 1, reused 0, downloaded 0, added 0, done
Done in 407ms using pnpm v11.22.0

$ cat package.json
{
        "dependencies": {
                "typescript": "^6.0.3"
        }
}

$ pnpm add --save-exact typescript@6.0.2
...
Packages: +1 -1
+-
Progress: resolved 1, reused 0, downloaded 1, added 1, done

dependencies:
- typescript 6.0.3
+ typescript 6.0.2 (7.0.2 is available)

Done in 882ms using pnpm v11.22.0

$ cat package.json
{
        "dependencies": {
                "typescript": "^6.0.2"
        }
}
```

Seems like unusual behavior that version ranges are not overwritten with [`--save-exact`](https://pnpm.io/cli/add#--save-exact--e--e).

Would a PR be accepted for this?

**Comment by btea** [comment]
The `pnpm up` command does not support the `--save-exact` parameter, whereas the `pnpm add` command does.

**Comment by karlhorky** [comment]
@btea oh yes, I see the original issue reports `pnpm up`. @unional can you change the issue title and description to `pnpm add` instead?

If you check out [my comment](https://github.com/pnpm/pnpm/issues/6398#issuecomment-5314388945) you'll see there is indeed an issue with `pnpm add --save-exact` as well though - the version range stays across multiple `pnpm add --save-exact` commands.

This unexpected behavior is also reported over in issue https://github.com/pnpm/pnpm/issues/8300

**Comment by btea** [comment]
https://github.com/pnpm/pnpm/issues/6398#issuecomment-5314388945 Based on your example above, the `save-extract` parameter appears to work correctly with the `pnpm add` command. 

When the `save-extract` parameter is enabled, dependencies are installed using explicitly defined versions rather than pnpm's default semver range operator.

**Comment by karlhorky** [comment]
@btea Hm, I don't understand what you mean - I thought that the example in [my comment above](https://github.com/pnpm/pnpm/issues/6398#issuecomment-5314388945) was clearly showing unusual behavior.

Here's a smaller version:

My understanding is that [`pnpm add --save-exact`](https://pnpm.io/cli/add#--save-exact--e--e) would switch from the `^6.0.3` version range to `6.0.2` - but it instead yields `^6.0.2` (version range with the `^` caret):

```bash
$ cat package.json
{
        "dependencies": {
                "typescript": "^6.0.3"
        }
}

$ pnpm add --save-exact typescript@6.0.2
...
Packages: +1 -1
+-
Progress: resolved 1, reused 0, downloaded 1, added 1, done

dependencies:
- typescript 6.0.3
+ typescript 6.0.2 (7.0.2 is available)

Done in 882ms using pnpm v11.22.0

$ cat package.json
{
        "dependencies": {
                "typescript": "^6.0.2"
        }
}
```

Expected `package.json` at end (no `^` caret range):

```json
{
        "dependencies": {
                "typescript": "6.0.2"
        }
}
```

**Comment by btea** [comment]
> Expected `package.json` at end (no carat `^` range):
> 
> {
>         "dependencies": {
>                 "typescript": "6.0.2"
>         }
> }

As I understand it, `save-exact` does not change the version prefix; it simply ignores the prefix during the installation process and installs based on the explicit version number.

**Comment by karlhorky** [comment]
> `save-exact` does not change the version prefix

Right, that's what I believe the bug is, both here and in issue #8300 - because [the docs](https://pnpm.io/cli/add#--save-exact--e--e) (and the `--save-exact` name) appear to say the opposite:

> ### --save-exact, -E, -e
>
> Saved dependencies will be configured with an exact version rather than using pnpm's default semver range operator.

I would expect this exact version to not only appear in the lockfile, but also in `package.json`.

**Comment by btea** [comment]
Oh, so this was a bug? I always thought it was expected behavior. :man_facepalming:

**Comment by karlhorky** [comment]
@zkochan what are your thoughts?

Given this `package.json`:

```json
{
        "dependencies": {
                "typescript": "^6.0.3"
        }
}
```

Should `pnpm add --save-exact typescript@6.0.2`:

1. Current behavior: update `package.json` to a new range `^6.0.2` (seems confusing / unexpected to me)
2. New behavior: update `package.json` to pin to `6.0.2`

If pnpm should create a new range `^6.0.2` then that makes it hard for users to switch deps in `package.json` from a range to a pinned exact version. (eg. needs first `pnpm rm typescript` and then `pnpm add --save-exact typescript@6.0.2`)

**Comment by zkochan** [comment]
Closing as a duplicate of #5436 (2022-10-01), which is the older report of the same thing.

Same bug: `--save-exact` is ignored by `pnpm update` and `pnpm up`, which keep the range prefix instead of writing the exact version.

Follow #5436 for this. If you think the two differ in a way that matters, say so here and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
